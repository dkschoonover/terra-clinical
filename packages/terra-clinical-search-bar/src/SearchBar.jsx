import React, { PropTypes } from 'react';
import Button from 'terra-button';
import IconSearch from 'terra-icon/lib/icon/IconSearch';
import './SearchBar.scss';

const propTypes = {
  /**
   * Placeholder text to show while the search field is empty.
   */
  placeholder: PropTypes.string,

  /**
   * The minimum number of characters to perform a search.
   */
  minimumSearchTextLength: PropTypes.number,

  /**
   * How long the component should wait after input before performing an automatic search.
   */
  searchDelay: PropTypes.number,

  /**
   * A callback to perform search. Sends parameter {String} searchText.
   */
  onSearch: PropTypes.func,
};

const defaultProps = {
  placeholder: '',
  minimumSearchTextLength: 2,
  searchDelay: 250,
};

class SearchBar extends React.Component {

  constructor(props) {
    super(props);

    this.handleTextChange = this.handleTextChange.bind(this);
    this.handleSearch = this.handleSearch.bind(this);

    this.searchTimeout = null;
    this.state = {
      searchText: '',
    };
  }

  handleTextChange(event) {
    this.setState({ searchText: event.target.value });

    this.clearSearchTimeout();
    this.searchTimeout = setTimeout(this.handleSearch, this.props.searchDelay);
  }

  handleSearch() {
    if (this.state.searchText.length >= this.props.minimumSearchTextLength) {
      this.clearSearchTimeout();

      if (this.props.onSearch) {
        this.props.onSearch(this.state.searchText);
      }
    }
  }

  clearSearchTimeout() {
    if (this.searchTimeout) {
      clearTimeout(this.searchTimeout);
      this.searchTimeout = null;
    }
  }

  render() {
    return (
      <div className="terraClinical-SearchBar">
        <input className="terraClinical-SearchBar-input" type="search" placeholder={this.props.placeholder} value={this.state.searchText} onChange={this.handleTextChange} />
        <Button className="terraClinical-SearchBar-button" onClick={this.handleSearch}>
          <IconSearch />
        </Button>
      </div>
    );
  }

}

SearchBar.propTypes = propTypes;
SearchBar.defaultProps = defaultProps;

export default SearchBar;
