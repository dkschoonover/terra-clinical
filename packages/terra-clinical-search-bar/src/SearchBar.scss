@import '~terra-mixins';

.terraClinical-SearchBar {
  background-color: $terra-grey-5;
  display: inline-flex;
  justify-content: start;
  padding: 0.714rem;
  width: 100%;

  .terraClinical-SearchBar-input {
    @include terra-border-radius-start(0.286em);
    @include terra-border-radius-end(0);

    appearance: textfield;
    background-color: $terra-white;
    border: 1px solid $terra-grey-15;
    border-right: 0.5px;
    flex: 1 1 auto;
    font-family: inherit;
    font-size: 16px;
    line-height: 1.4285714285714286rem;
    margin: 0;

    // This is needed when the search bar is used in another wrapper that uses flexbox so that the search bar doesn't
    //  overlap with it's sibling in firefox.
    min-width: 0;

    padding: 0.286rem;
    width: 100%;
  }

  .terraClinical-SearchBar-button {
    @include terra-border-radius-start(0);
    @include terra-border-radius-end(0.286em);

    border: 1px solid $terra-grey-15;
    display: inline-block;
    flex: 0 0 auto;
    margin: 0;
  }
}

.terraClinical-SearchBar-input::-webkit-search-cancel-button {
  appearance: searchfield-cancel-button !important;
}

.terraClinical-SearchBar-input::-webkit-search-decoration {
  // Get rid of extra left-padding getting added in Safari
  appearance: none;
}
