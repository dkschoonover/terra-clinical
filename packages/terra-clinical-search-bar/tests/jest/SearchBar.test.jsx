import React from 'react';
import SearchBar from '../../src/SearchBar';


describe('Snapshots', () => {
  it('renders a basic search bar', () => {
    const searchBar = shallow(<SearchBar />);
    expect(searchBar).toMatchSnapshot();
  });

  it('renders a search bar with a placeholder', () => {
    const searchBar = shallow(<SearchBar placeholder="Test" />);
    expect(searchBar).toMatchSnapshot();
  });
});

describe('Manual Search', () => {
  it('triggers search on button press', () => {
    const onSearch = jest.fn();
    const searchBar = shallow(<SearchBar onSearch={onSearch} />);
    searchBar.setState({ searchText: 'Te' });

    expect(onSearch).not.toBeCalled();
    searchBar.childAt(1).simulate('click');
    expect(onSearch).toBeCalledWith('Te');
  });

  it('does not trigger search if default minimum search text has not been met', () => {
    const onSearch = jest.fn();
    const searchBar = shallow(<SearchBar onSearch={onSearch} />);
    searchBar.setState({ searchText: 'T' });

    expect(onSearch).not.toBeCalled();
    searchBar.childAt(1).simulate('click');
    expect(onSearch).not.toBeCalled();
  });

  it('does not trigger search if minimum search text has not been met', () => {
    const onSearch = jest.fn();
    const searchBar = shallow(<SearchBar onSearch={onSearch} minimumSearchTextLength={5} />);
    searchBar.setState({ searchText: 'Sear' });

    expect(onSearch).not.toBeCalled();
    searchBar.childAt(1).simulate('click');
    expect(onSearch).not.toBeCalled();
  });

  it('does not search when callback is not provided', () => {
    const searchBar = shallow(<SearchBar minimumSearchTextLength={5} />);
    searchBar.setState({ searchText: 'Searc' });

    searchBar.childAt(1).simulate('click'); // Verifies we do not attempt to call an undefined function.
  });
});

describe('Auto Search', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('triggers search after delay on text change', () => {
    const onSearch = jest.fn();
    const searchBar = shallow(<SearchBar onSearch={onSearch} />);

    searchBar.childAt(0).simulate('change', { target: { value: 'Te' } });

    expect(onSearch).not.toBeCalled();
    jest.runAllTimers();
    expect(onSearch).toBeCalledWith('Te');
  });

  it('does not trigger search if minimum text length is not met', () => {
    jest.useFakeTimers();
    const onSearch = jest.fn();
    const searchBar = shallow(<SearchBar onSearch={onSearch} minimumSearchTextLength={5} />);

    searchBar.childAt(0).simulate('change', { target: { value: 'Sear' } });

    expect(onSearch).not.toBeCalled();
    jest.runAllTimers();
    expect(onSearch).not.toBeCalled();
  });

  it('uses standard timeout for search delay when not provided', () => {
    const searchBar = shallow(<SearchBar />);

    searchBar.childAt(0).simulate('change', { target: {} });
    expect(setTimeout).toBeCalledWith(expect.anything(), 250);
  });

  it('uses custom timeout for search delay when provided', () => {
    const searchBar = shallow(<SearchBar searchDelay={1000} />);

    searchBar.childAt(0).simulate('change', { target: {} });
    expect(setTimeout).toBeCalledWith(expect.anything(), 1000);
  });
});
