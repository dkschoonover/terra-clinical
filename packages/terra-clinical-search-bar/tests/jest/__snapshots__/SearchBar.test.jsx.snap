exports[`Snapshots renders a basic search bar 1`] = `
<div
  className="terraClinical-SearchBar">
  <input
    className="terraClinical-SearchBar-input"
    onChange={[Function]}
    placeholder=""
    type="search"
    value="" />
  <Button
    className="terraClinical-SearchBar-button"
    isBlock={false}
    isCompact={false}
    isDisabled={false}
    isReversed={false}
    onClick={[Function]}
    variant="default">
    <IconSearch
      data-name="Layer 1"
      viewBox="0 0 48 48"
      xmlns="http://www.w3.org/2000/svg" />
  </Button>
</div>
`;

exports[`Snapshots renders a search bar with a placeholder 1`] = `
<div
  className="terraClinical-SearchBar">
  <input
    className="terraClinical-SearchBar-input"
    onChange={[Function]}
    placeholder="Test"
    type="search"
    value="" />
  <Button
    className="terraClinical-SearchBar-button"
    isBlock={false}
    isCompact={false}
    isDisabled={false}
    isReversed={false}
    onClick={[Function]}
    variant="default">
    <IconSearch
      data-name="Layer 1"
      viewBox="0 0 48 48"
      xmlns="http://www.w3.org/2000/svg" />
  </Button>
</div>
`;
